src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2497,33 +2497,41 @@ protected virtual MethodDefinition MarkMethod (MethodReference reference, Depend
 			return method;
 		}
 
-		protected internal void ProcessRequiresUnreferencedCode (MethodDefinition method, MessageOrigin origin, DependencyKind dependencyKind)
+		void ProcessRequiresUnreferencedCode (MethodDefinition method, MessageOrigin origin, DependencyKind dependencyKind)
 		{
 			switch (dependencyKind) {
 			case DependencyKind.AccessedViaReflection:
-			case DependencyKind.DirectCall:
 			case DependencyKind.DynamicallyAccessedMember:
 			case DependencyKind.DynamicDependency:
 			case DependencyKind.ElementMethod:
 			case DependencyKind.Ldftn:
 			case DependencyKind.Ldvirtftn:
-			case DependencyKind.Newobj:
 			case DependencyKind.TriggersCctorForCalledMethod:
-			case DependencyKind.VirtualCall:
 				break;
 
+			// DirectCall, VirtualCall and NewObj are handled by ReflectionMethodBodyScanner
+			// This is necessary since the ReflectionMethodBodyScanner has intrinsic handling for some
+			// of the methods annotated with the attribute (for example Type.GetType)
+			// and it know when it's OK and when it needs a warning. In this place we don't know
+			// and would have to warn every time.
+
```

Comment:
Is it worth adding assert for that?

Reply:
It's very hard to "verify" this - the code will get here for DirectCall (and the others), actually before it gets to `ReflectionMethodBodyScanner`. The only sensible check would be in `ProcessMethodBody` that if we see a method call to annotated method, we enable `ReflectionMethodBodyScanner` - but that is already there, since that's the place which enables the data flow processing.

This code here runs during `MarkMethod`, the data flow analysis runs during `ProcessMethod`, so much later on.

I do agree that this relationship is a bit fragile, but I can't think of a better way to handle it right now.